### ISSUE TITLE
Ability to preload modules when workers start

### ISSUE THREAD

**Issue by nirvana-msu** [ray-project/ray#13908]
### Describe your feature request

In any reasonably large project you would face an issue of circular imports. This is often handled by having a pre-defined entry point - a module that always loads first, bootstrapping the rest of the application. For instance, in web applications that would be the module holding WSGI / ASGI app. You can easily get a circular import error if the first module you attempt to import is some other random module, and not the intended one.

This is exactly what happens within Ray workers.  E.g. when I start an Actor worker and it tries to unpickle the actor class, the first module it attempts to load is the module where actor is defined - leading to circular import errors as described above.

Just in case, I'm aware of  `_code_search_path` parameter of `ray.init`, but that doesn't help. All my modules are already on PATH, the only thing that's wrong is the order in which they get imported in a Ray worker.

There should be an option to provide a list of modules, that would be imported in a given order in each worker before any other user code is run. That would allow users to ensure correct module loading order - e.g. in my case I would specify the WSGI entry point as the module to preload.

For reference, exact same functionality exists (and is widely used) in all other distributed computing libraries, e.g.:
 * Dask has [preload](https://docs.dask.org/en/latest/setup/custom-startup.html#preload-scripts) option:
     > A module or Python file passed as a --preload value is guaranteed to be imported before establishing any connection.
 * Celery has [imports](https://docs.celeryproject.org/en/stable/userguide/configuration.html#imports) option:
     > A sequence of modules to import when the worker starts.
     > The modules will be imported in the original order.

I'm a bit surprised I can't seem to find any reference to this functionality in Ray - not even within issues/discussions. It would expect the absence of that option to cause a fair amount of headache to many reasonably large projects.

**Comment by rkooo567** [comment]
@richardliaw Do we have any way to achieve this now?

**Comment by mdwint** [comment]
I'm looking for this feature as well. So far I've tried `ray.worker.global_worker.run_function_on_all_workers`, but this doesn't work as expected. For some reason the function isn't run on all workers. In #12599 it was mentioned that this function is not a part of the public API, which seems strange for such a core feature.

**Comment by rkooo567** [comment]
@ericl Isn't this something we can resolve with the new runtime_env feature? Can you verify that?

**Comment by ericl** [comment]
cc @edoakes I think we'd still need some entrypoint hook right?

**Comment by edoakes** [comment]
@ericl yeah this will need something a bit separate. It should be easy enough to implement though. We can stuff it in the runtime_env and just do those imports before anything else starts up.

**Comment by edoakes** [comment]
@nirvana-msu I think it's be simplest to support something similar to celery here -- are there downsides to this / cases it doesn't cover when compared to dask's preload?

**Comment by nirvana-msu** [comment]
@edoakes Dask config option seems overly-complicated to my taste (as is the most of Dask API... so refreshing to use Ray in that sense). Celery-style list of modules to import seems perfectly suitable to me.

**Comment by edoakes** [comment]
@nirvana-msu Sounds good, I agree. We will add this to the backlog and address it sometime soon as part of our existing `runtime_env` work.

**Comment by nikunjgit** [comment]
Any updates here?

**Comment by wwoods** [comment]
Also would like to use this to avoid pickle errors. However, sometimes I need to run a dynamic import directly via `importlib.machinery.SourceFileLoader` -- so a static list of imports that just get enumerated on the worker may not suffice. Some sort of setup function would be a lot more flexible.

**Comment by nirvana-msu** [comment]
You can put any code you want in a module that gets imported, so a static list of modules still covers this I would have thought. That is unless you want to pass some parameters to it.

**Comment by wwoods** [comment]
> You can put any code you want in a module that gets imported, so a static list of modules still covers this I would have thought. That is unless you want to pass some parameters to it.

Fair point. Parameters are involved, though realistically that might just be bad code hygiene.
